Add YouTube Video Playback Support to Movie & Series Details

skaik-mo opened the pull request:
This PR introduces full support for playing YouTube trailers directly from the movie and series details screens. It includes the integration of the YouTube Android Player library, updates to data models and repositories to fetch trailer information, and UI enhancements for navigating and displaying trailers.


  Key Features

   - YouTube Video ID Support:
    Added youtubeVideoId to movie and series domain/UI models, DTOs, and mappers.

   - Trailer API Integration:
    Implemented API calls to fetch YouTube trailer keys for movies and series. Prioritizes "Trailer" type.

   - YouTube Player Screen:
    Created YouTubePlayerScreen.kt using androidyoutubeplayer. Added navigation and composable integration.

   - Play Button Behavior:
    Play button now navigates to trailer screen if youtubeVideoId exists; otherwise, it is disabled with dimmed UI.

   - Library & Navigation Setup:
    Added required library dependency and updated navigation graph for YouTube player route.


https://github.com/user-attachments/assets/9765cefe-c260-445d-a391-cfff3b760562

skaik-mo committed:
get Movie Trailer Url

-   Makes `MoviesListResponse` generic to support different types of results.
-   Adds `TrailerResponse` data class to represent movie trailer information.
-   Adds a `VIDEOS_END_POINT` constant to `NetworkConstants`.
-   Modifies `MoviesApiServiceRetrofit` to use the generic `MoviesListResponse` and adds a method to fetch movie trailer URLs.
-   Updates `MoviesRemoteDataSource` and `MoviesRemoteDataSourceImplRetrofit` to include the `getMovieTrailerUrl` method, which fetches and returns the key of the first trailer found.

skaik-mo committed:
get Series Trailer Url

- Genericizes `SeriesResponse` and `MoviesListResponse` to accept a type parameter `T` for the `results` list, allowing them to be reused for different data types.
- Makes `page`, `totalPages`, and `totalResults` in `SeriesResponse` and `MoviesListResponse` nullable and provides default `null` values.
- Removes the unused `id` field from `MoviesListResponse`.
- Moves `TrailerResponse` from the `com.giraffe.media.movie.response` package to `com.giraffe.media.response` to make it more generic.
- Adds `getSeriesTrailerUrl` function to `SeriesApiServiceRetrofit`, `SeriesRemoteRetrofitDataSourceImp`, and `SeriesRemoteDataSource` to fetch the trailer URL for a series.
- Implements `getSeriesTrailerUrl` in `SeriesRemoteRetrofitDataSourceImp` to retrieve the trailer key, prioritizing "Trailer" type videos and falling back to the first available video.

skaik-mo committed:
Add Android YouTube Player library and VideoPlayer composable

- Adds the `com.pierfrancescosoffritti.androidyoutubeplayer:core` dependency to the `presentation/details` module.
- Creates a new `VideoPlayer.kt` file with a composable function `YouTubePlayer`.
- The `YouTubePlayer` composable uses `AndroidView` to embed a `YouTubePlayerView`.
- It initializes the player with a given `youtubeVideoId` and includes an `AppBar` with a back button.

skaik-mo committed:
Add YouTube video player functionality

Key changes:
- Added `YouTubePlayerScreen.kt` to display and control the YouTube player using the `com.pierfrancescosoffritti.androidyoutubeplayer.core` library.
- Integrated `YouTubePlayerRoute.kt` for navigation to the `YouTubePlayerScreen`, passing the `youtubeVideoId`.
- Updated `MovieDetailsScreen.kt` and `SeriesDetailsScreen.kt` to navigate to the `YouTubePlayerScreen` when the play button is clicked, passing the `youtubeVideoId`.
- Modified `MainMovieOrSeriesDetailsAnimatedContent.kt`, `MinimizedInfoRow.kt`, and `MainMovieOrSeriesDetails.kt` to:
    - Accept an `isPlayButtonEnabled` parameter.
    - Change the play button's background color based on `isPlayButtonEnabled`.
    - Disable the play button's click functionality when `isPlayButtonEnabled` is false.
- Added `youtubeVideoId` to `MovieUi.kt` and `SeriesUi.kt` data models.
- Updated `DetailsNavGraph.kt` to include the `youTubePlayerRouteRoute`.
- The play button in movie/series details now triggers navigation to the YouTube player, passing the respective `youtubeVideoId`.
- The play button is disabled if `youtubeVideoId` is blank.

skaik-mo committed:
Add YouTube video ID to movie and series details

The following changes were made:

- Added `youtubeVideoId` to `Movie` and `Series` entities.
- Added `youtubeVideoId` to `MovieDto` and `SeriesDetailsDto`.
- Updated mappers to include `youtubeVideoId`.
- Modified `MoviesRepositoryImpl` and `SeriesRepositoryImpl` to fetch and store the `youtubeVideoId` when retrieving movie or series details.
- Updated `MovieUi` and `SeriesUi` to include `youtubeVideoId`.
- Updated unit tests to reflect the new field.

skaik-mo committed:
Merge branch 'develop' into feature/display-trailer

* develop: (44 commits)
  fix: - remove click action from PrimaryButton and SecondaryButton
  enhance arrow back ripple
  add dagger in view model
  feat: Update SearchResultViewModel to use state for query parameter
  Pass click event for create collection button to bottom sheet
  feat: Add onClickCreateCollection to SeriesDetails
  Fix padding issue in BaseBottomSheet
  Adjust BaseBottomSheet navigation bar padding
  feat: Remove Hilt-related metadata from AndroidManifest.xml
  feat: Remove Hilt-related metadata from AndroidManifest.xml
  enhancement false padding
  enhancement clickable items
  feat: Remove unused imports in MovieDetailsScreen.kt
  feat: Refactor APIs to use Hilt for dependency injection
  feat: Integrate Hilt for dependency injection in authentication modules
  feat: Implement Hilt dependency injection in authentication data sources
  remove unnecessary effect in the screen
  fix: change test cases
  Update presentation/authentication/src/main/java/com/giraffe/authentication/screen/LoginViewModel.kt
  Adjust MovieDetailsScreen collection bottom sheet padding
  ...

# Conflicts:
#	gradle/libs.versions.toml
#	presentation/details/src/main/java/com/giraffe/details/screens/moviedetails/screen/MovieDetailsRoute.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/moviedetails/screen/MovieDetailsScreen.kt
#	repository/media/src/main/java/com/giraffe/media/series/SeriesRepositoryImpl.kt

mohamedebrahem13 commented:
i think it should navigate to YouTube not show the video in our app @skaik-mo

Ashraf-El-Fallah commented on datasource/remote/media/src/main/java/com/giraffe/media/response/TrailerResponse.kt:
```diff
@@ -0,0 +1,22 @@
+    var type: String? = null,
+    var official: Boolean? = null,
+    @SerialName("iso_639_1")
+    var iso6391: String? = null,
```
You can change the name of variable if you use @SerialName

Ashraf-El-Fallah commented on presentation/details/src/main/java/com/giraffe/details/nav/DetailsNavGraph.kt:
```diff
@@ -78,6 +81,9 @@ internal fun DetailsNavGraph(
+        youTubePlayerRouteRoute(
+            onBackClick = { navController.navigateUp() }
+        )
 //        reviewRoute(navController)
```
remove commented code if you don't need it

Mostafa-khaled21 commented:
> i think it should navigate to YouTube not show the video in our app @skaik-mo

I don't think so. See Ahmed Mounes's issue for Cairo Squad. @mohamedebrahem13

skaik-mo committed:
Merge branch 'develop' into feature/display-trailer

* develop: (37 commits)
  Apply navigationBarsPadding to BottomNavigationBar
  Refactor: Improve error handling and data loading in ViewModels
  fix: Update SafeIslamicImageClassifier to use two NSFW models and simplify classification logic
  Translate Explore screen strings to Arabic
  Refactor: Improve data fetching and state management in ViewModels
  add padding for system bars
  add animated in bottom navigation bar
  Refactor: Remove key from LazyRow and LazyVerticalGrid items
  Apply clickable modifier to entire poster card
  Make data class internal in module
  remove unnecessary animation from explore nav graph
  add temporal match and profile screens
  remove wrong animation from bottom nav bar
  add match and profile tab items to the bottom nav bar
  add profile and match screens to the nav graph
  add profile and match tabs
  add profile api
  add match api
  Refactor: Update recent series retrieval to use Flow
  add string resources
  ...

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/components/MainMovieOrSeriesDetailsAnimatedContent.kt
#	presentation/details/src/main/java/com/giraffe/details/models/MovieUi.kt
#	presentation/details/src/main/java/com/giraffe/details/nav/DetailsNavGraph.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/screen/SeriesDetailsRoute.kt

skaik-mo committed:
Remove iso6391 and iso31661 fields from TrailerResponse

The fields `iso6391` and `iso31661` were removed from the `TrailerResponse` data class as they were not being used.

skaik-mo committed:
Merge branch 'develop' into feature/display-trailer

* develop:
  fix home view model and movie list view model
  Refactor: Simplify SeriesDetailsRoute navigation
  add home nav graph from develop
  Remove Reviews screen navigation from series details
  Translate strings to Arabic
  Test: Add unit tests for IsLoggedInUseCase
  add content description
  use fixed grid cells in LazyHorizontalGrid
  remove not used map function in view model
  remove unused function
  using string resources instead of static value
  remove dead and unused code
  feat: Implement login prompt for rating and add to collection
  use enum for static collection images
  resolve film and series don't be arabic
  resolve arrow direction in arabic
  fix base view model
  resolve dagger in view model
  resolve conflicts in develop
  make collection navigate to list screen and display the movies related to collection

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/nav/DetailsNavGraph.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/moviedetails/screen/MovieDetailsRoute.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/screen/SeriesDetailsContent.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/screen/SeriesDetailsRoute.kt
#	presentation/details/src/main/java/com/giraffe/details/screens/seriesdetails/screen/SeriesDetailsScreen.kt

skaik-mo committed:
Refactor: Use `rememberSaveable` for `isBottomBarVisible`
